stan/math/prim/fun/binomial_coefficient_log.hpp

```diff
@@ -48,32 +57,91 @@ namespace math {
- * @param n number of objects chosen.
- * @return log (N choose n).
+ *  This function is numerically more stable than naive evaluation via lgamma.
+ *
+ * @tparam T_n type of the first argument
+ * @tparam T_k type of the second argument
+ *
+ * @param n total number of objects.
+ * @param k number of objects chosen.
+ * @return log (n choose k).
  */
-template <typename T_N, typename T_n>
-inline return_type_t<T_N, T_n> binomial_coefficient_log(const T_N N,
-                                                        const T_n n) {
-  const double CUTOFF = 1000;
-  if (N - n < CUTOFF) {
-    const T_N N_plus_1 = N + 1;
-    return lgamma(N_plus_1) - lgamma(n + 1) - lgamma(N_plus_1 - n);
+
+template <typename T_n, typename T_k>
+inline return_type_t<T_n, T_k> binomial_coefficient_log(const T_n n,
+                                                        const T_k k) {
+  if (is_any_nan(n, k)) {
+    return stan::math::NOT_A_NUMBER;
+  }
+
+  // Choosing the more stable of the symmetric branches
+  if (n > 0 && k > value_of_rec(n) / 2.0 + 1e-8) {
+    return binomial_coefficient_log(n, n - k);
+  }
+
+  using T_partials_return = partials_return_t<T_n, T_k>;
+
+  const T_partials_return n_ = value_of(n);
+  const T_partials_return k_ = value_of(k);
+  const T_partials_return n_plus_1 = n_ + 1;
+  const T_partials_return n_plus_1_mk = n_plus_1 - k_;
+
+  static const char* function = "binomial_coefficient_log";
+  check_greater_or_equal(function, "first argument", n, -1);
+  check_greater_or_equal(function, "second argument", k, -1);
+  check_greater_or_equal(function, "(first argument - second argument + 1)",
+                         n_plus_1_mk, 0.0);
+
+  operands_and_partials<T_n, T_k> ops_partials(n, k);
+
+  T_partials_return value;
+  if (k_ == 0) {
+    value = 0;
+  } else if (n_plus_1 < lgamma_stirling_diff_useful) {
+    value = lgamma(n_plus_1) - lgamma(k_ + 1) - lgamma(n_plus_1_mk);
   } else {
-    return_type_t<T_N, T_n> N_minus_n = N - n;
-    const double one_twelfth = inv(12);
-    return multiply_log(n, N_minus_n) + multiply_log((N + 0.5), N / N_minus_n)
-           + one_twelfth / N - n - one_twelfth / N_minus_n - lgamma(n + 1);
+    value = -lbeta(n_plus_1_mk, k_ + 1) - log1p(n_);
   }
+
+  if (!is_constant_all<T_n, T_k>::value) {
+    // Branching on all the edge cases.
+    // In direct computation many of those would be NaN
+    // But one-sided limits from within the domain exist.
+    T_partials_return digamma_n_plus_1_mk = digamma(n_plus_1_mk);
+
+    if (!is_constant_all<T_n>::value) {
+      if (n_ == -1.0) {
+        if (k_ == 0) {
+          ops_partials.edge1_.partials_[0] = 0;
+        } else {
+          ops_partials.edge1_.partials_[0] = stan::math::NEGATIVE_INFTY;
+        }
+      } else {
+        ops_partials.edge1_.partials_[0]
+            = (digamma(n_plus_1) - digamma_n_plus_1_mk);
+      }
+    }
+    if (!is_constant_all<T_k>::value) {
+      if (k_ == 0 && n_ == -1.0) {
```

Comment:
Could you explain the gradient logic for k here?

I think I kinda followed the ones for n but these ones confused me.

Reply:
Not sure what to add. The only fact I use is that `lim x-> 0 digamma(x)` from above is negative infinity. Mentioned that in the comments, if that is still confusing, let me know (it is also possible I am missing some of the edge cases).

Reply:
Ah yeah, that makes sense.